k8s, cluster-up: Copy istio CNI net conf to where Multus expects

ormergi opened the pull request:
Since Multus v4 it expects CNI net conf file to exist at `/etc/cni/net.d` but Isitio generates its net conf file at `/etc/cni/multus/net.d`.

Copy istio CNI net conf to where multus expects on cluster-up in order to enable deploying both Isitio and Multus until proper solution is available, tracking issue [1] [2].

[1] https://github.com/kubevirt/kubevirtci/issues/906 [2] https://github.com/k8snetworkplumbingwg/multus-cni/issues/982

Signed-off-by: Or Mergi <ormergi@redhat.com>

ormergi commented:
/cc @maiqueb @brianmcarey @dhiller

phoracek commented on cluster-up/cluster/k8s-provider-common.sh:
```diff
@@ -88,6 +88,14 @@ function wait_for_istio_ready() {
+function copy_istio_cni_conf_files() {
+  if [ "$KUBEVIRT_DEPLOY_ISTIO" == "true" ] && [ "$KUBEVIRT_WITH_CNAO" == "true" ]; then
+    for nodeNum in $(seq -f "%02g" 1 $KUBEVIRT_NUM_NODES); do
+      $ssh node${nodeNum} -- sudo cp -uv /etc/cni/multus/net.d/*istio*.conf /etc/cni/net.d/
```
Would creating a symlink be an option too? It would allow us to keep it generic and we would not need to wait for Istio to finish deployment

ormergi replied:
@phoracek  it requires knowing the name of isito conf file that will be generated in order to create the link (though it won't likely to change), then it would be possible to create the link as part of `deploy_istio()` (before Isitio deployment ready), for example:
```
ISTIO_CONF_FILE_NAME="YYY-istio-cni.conf"
...
function deploy_istio() {
    if [ "$KUBEVIRT_DEPLOY_ISTIO" == "true" ]; then
        if [ "$KUBEVIRT_WITH_CNAO" == "true" ]; then
        ...
        $ssh node${nodeNum} -- sudo ln -sr /etc/cni/multus/net.d/${ISTIO_CONF_FILE_NAME} /etc/cni/net.d/
...
```
WDYT?

ormergi replied:
It wont work because Multus dosent look for net conf files recursively nor look into links, the conf file must be copied.

phoracek replied:
i see, interesting, thanks for checking that

dhiller commented on cluster-up/cluster/k8s-provider-common.sh:
```diff
@@ -88,6 +88,14 @@ function wait_for_istio_ready() {
 }
 
+function copy_istio_cni_conf_files() {
+  if [ "$KUBEVIRT_DEPLOY_ISTIO" == "true" ] && [ "$KUBEVIRT_WITH_CNAO" == "true" ]; then
```
nit: please adjust indentation to the remainder of the script

ormergi replied:
Thanks for the heads up, done

ormergi commented on cluster-up/cluster/k8s-provider-common.sh:
```diff
@@ -155,4 +163,5 @@ function up() {
         sleep 5
     done
 
+    copy_istio_cni_conf_files
```
Done

ormergi commented:
@phoracek @dhiller I have resolved your review comments please take another look :slightly_smiling_face:

phoracek reviewed (approved):
Thanks

ormergi commented:
[Changes](https://github.com/kubevirt/kubevirtci/compare/ee2e69fb1ea9764b42ca9dfbf5ea1e787a30b6dd..af74d5241f9035953eec20c84340b5fad675ca2c): fixed missing if closing, missing indentation and removed the fixme comment leftover

maiqueb reviewed (approved):
Thanks.

maiqueb commented:
Weird, the 1.26 provisioning lane is failing w/
```
+ dnf install -y kernel-modules-5.14.0-210.el9.x86_64
CentOS Stream 9 - BaseOS                        8.8 MB/s | 6.0 MB     00:00    
CentOS Stream 9 - AppStream                      10 MB/s |  16 MB     00:01    
CentOS Stream 9 - Extras packages                11 kB/s | 9.2 kB     00:00    
No match for argument: kernel-modules-5.14.0-210.el9.x86_64
Error: Unable to find a match: kernel-modules-5.14.0-210.el9.x86_64
```

But I see the pkg available online:
http://mirror.stream.centos.org/9-stream/BaseOS/x86_64/os/Packages/kernel-modules-5.14.0-210.el9.x86_64.rpm

... let's blindly retest once.
/test check-provision-k8s-1.26

maiqueb commented:
Can we be missing some packages on the 1.26 lanes ? 
@dhiller / @brianmcarey 
```
podman run --rm --entrypoint bash -it centos:stream9 
Resolved "centos" as an alias (/etc/containers/registries.conf.d/000-shortnames.conf)
Trying to pull quay.io/centos/centos:stream9...
Getting image source signatures
Copying blob 5af5d4bc82b8 done  
Copying config bc5c131925 done  
Writing manifest to image destination
Storing signatures
[root@95b437f4fede /]# dnf info --showduplicates kernel-modules | grep Version -A1
Version      : 5.14.0
Release      : 196.el9
--
Version      : 5.14.0
Release      : 197.el9
--
Version      : 5.14.0
Release      : 200.el9
--
Version      : 5.14.0
Release      : 202.el9
--
Version      : 5.14.0
Release      : 205.el9
```

maiqueb commented:
/test check-provision-k8s-1.25-ipv6

dhiller commented:
> Can we be missing some packages on the 1.26 lanes ? @dhiller / @brianmcarey
> 
> ```
> podman run --rm --entrypoint bash -it centos:stream9 
> Resolved "centos" as an alias (/etc/containers/registries.conf.d/000-shortnames.conf)
> Trying to pull quay.io/centos/centos:stream9...
> Getting image source signatures
> Copying blob 5af5d4bc82b8 done  
> Copying config bc5c131925 done  
> Writing manifest to image destination
> Storing signatures
> [root@95b437f4fede /]# dnf info --showduplicates kernel-modules | grep Version -A1
> Version      : 5.14.0
> Release      : 196.el9
> --
> Version      : 5.14.0
> Release      : 197.el9
> --
> Version      : 5.14.0
> Release      : 200.el9
> --
> Version      : 5.14.0
> Release      : 202.el9
> --
> Version      : 5.14.0
> Release      : 205.el9
> ```

That's strange. I see a kernel version change from one job run to the other:

```
```
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirtci/932/check-provision-k8s-1.26/1602565136952856576#1:build-log.txt%3A2309
dnf install -y kernel-modules-5.14.0-205.el9.x86_64
```

```
dnf install -y kernel-modules-5.14.0-210.el9.x86_64
```
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirtci/932/check-provision-k8s-1.26/1602676155033325568#1:build-log.txt%3A2311

even though the box version is `Centos9 version **20221129.1` in both runs

maiqueb commented:
> > Can we be missing some packages on the 1.26 lanes ? @dhiller / @brianmcarey
> > ```
> > podman run --rm --entrypoint bash -it centos:stream9 
> > Resolved "centos" as an alias (/etc/containers/registries.conf.d/000-shortnames.conf)
> > Trying to pull quay.io/centos/centos:stream9...
> > Getting image source signatures
> > Copying blob 5af5d4bc82b8 done  
> > Copying config bc5c131925 done  
> > Writing manifest to image destination
> > Storing signatures
> > [root@95b437f4fede /]# dnf info --showduplicates kernel-modules | grep Version -A1
> > Version      : 5.14.0
> > Release      : 196.el9
> > --
> > Version      : 5.14.0
> > Release      : 197.el9
> > --
> > Version      : 5.14.0
> > Release      : 200.el9
> > --
> > Version      : 5.14.0
> > Release      : 202.el9
> > --
> > Version      : 5.14.0
> > Release      : 205.el9
> > ```
> 
> That's strange. I see a kernel version change from one job run to the other:
> 
> ```
> dnf install -y kernel-modules-5.14.0-205.el9.x86_64
> ```
> 
> https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirtci/932/check-provision-k8s-1.26/1602565136952856576#1:build-log.txt%3A2309
> 
> ```
> dnf install -y kernel-modules-5.14.0-210.el9.x86_64
> ```
> 
> https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirtci/932/check-provision-k8s-1.26/1602676155033325568#1:build-log.txt%3A2311
> 
> even though the box version is `Centos9 version 20221129.1` in both runs

This is installed via `dnf install -y "kernel-modules-$(uname -r)" ` . [Code reference](https://github.com/kubevirt/kubevirtci/blob/main/cluster-provision/k8s/1.26/provision.sh#L88)

Where will this script be run ?... on the baremetal CI node ?... Could it be running this in different nodes will yield different packages to be installed ?...

dhiller commented:
So, I don't see any changes to the box on this page: https://cloud.centos.org/centos/9-stream/x86_64/images/
(It says the last change was done on 30th Nov. We might try to use a newer version of the Centos 9 box from that page.

dhiller commented:
I am trying https://cloud.centos.org/centos/9-stream/x86_64/images/CentOS-Stream-Vagrant-9-20221206.0.x86_64.vagrant-libvirt.box

dhiller commented:
> I am trying https://cloud.centos.org/centos/9-stream/x86_64/images/CentOS-Stream-Vagrant-9-20221206.0.x86_64.vagrant-libvirt.box

aaaand, it ofcourse has the same issue :confused:

ormergi commented:
Please note that this PR change has nothing to do with `check-provision-k8s-1.26` and `
check-up-kind-1.23-vgpu` lane failures :slightly_smiling_face:

dhiller commented:
> Please note that this PR change has nothing to do with `check-provision-k8s-1.26` and ` check-up-kind-1.23-vgpu` lane failures slightly_smiling_face

Yes but it will break the publishing, since the 1.26 will fail to build

ormergi commented:
> Yes but it will break the publishing, since the 1.26 will fail to build

Oh, I didn't realize that, I though that it wont be included in the publish process since the lane is not mandatory :/

Barakmor1 commented:
> During provisioning the [provision.sh](https://github.com/kubevirt/kubevirtci/blob/c1579907e89a5fbec9f7eab45d6287b11094ed3c/cluster-provision/k8s/1.26/provision.sh#L88-L87) is being run inside the Centos Stream 9 box, which is a vm that is started inside a [container image](https://github.com/kubevirt/kubevirtci/blob/71777d36c504aa74d0321a0c653847723426e645/cluster-provision/centos9/Dockerfile#L26-L25). This image should have the same kernel version for every run, but seemingly it changed. Asking @Barakmor1 to keep me honest here.

We change the Centos Stream 9 version is because the lane started failing few days ago:
https://github.com/kubevirt/kubevirtci/pull/924

brianmcarey commented:
/test check-provision-k8s-1.26

ormergi commented:
> I am suspecting that the change originates from a change in vmlinuz and/or initrd.img which are both pulled [during image creation](https://github.com/kubevirt/kubevirtci/blob/71777d36c504aa74d0321a0c653847723426e645/cluster-provision/centos9/Dockerfile#L32).
> 
> I am seeing a change on the initrd.img from the 12th of Dec and for vmlinuz on the 9th of Dec here: http://mirror.stream.centos.org/9-stream/BaseOS/x86_64/os/images/pxeboot/

Note that it is the same for centos8, `initd.img` and `vmlinuz` have different release dates 
http://mirror.centos.org/centos/8-stream/BaseOS/x86_64/os/images/pxeboot/

ormergi commented:
@dhiller @brianmcarey I think its good direction, base image and `initd.image` & `vmlinuz` are from different release date, it may explain why we see inconsistent outout of `uname -r` (as the version of `initd.image` & `vmlinuz` is not pinned).
Maybe it's worth testing it with a base images from the same release date as `initd.image` & `vmlinuz`.

Make sense?

brianmcarey commented:
/test check-provision-k8s-1.26

dhiller commented:
/test check-provision-k8s-1.26
/override check-up-kind-1.23-vgpu

dhiller commented:
/approve

brianmcarey commented:
/test check-provision-k8s-1.26

oshoval commented on cluster-up/cluster/k8s-provider-common.sh:
```diff
@@ -88,6 +88,16 @@ function wait_for_istio_ready() {
+# copy_istio_cni_conf_files copy the generated Istio CNI net conf file
+# (at '/etc/cni/multus/net.d/') to where Multus expect CNI net conf files ('/etc/cni/net.d/')
+function copy_istio_cni_conf_files() {
+    if [ "$KUBEVIRT_DEPLOY_ISTIO" == "true" ] && [ "$KUBEVIRT_WITH_CNAO" == "true" ]; then
```
what about `KUBVIRT_WITH_CNAO_SKIP_CONFIG` ?
The CR might be installed bit later and include the multus (but not sure if those folders do exists in this case at this stage ?)

can be ignored, if there is code that returns error upon this configuration

ormergi replied:
>what about KUBVIRT_WITH_CNAO_SKIP_CONFIG ?
The CR might be installed bit later and include the multus (but not sure if those folders do exists in this case at this stage ?)

If by CR you mean CNAO CR, in the scenrtio of Multus deployment hangs no harm will do to the cluster if Istio confs will be copied to `/etc/cni/net.d`

Note that `copy_istio_cni_conf_files` is called only after both CNAO and Istio deployments are ready, so it is not likely to happen.

> This function is called only after both CNAO and Istio deployments are ready, so it is not likely to happen.

I am not sure what you mean, the error flag is set for script.

oshoval replied:
I mean that the condition checks `KUBEVIRT_WITH_CNAO` but doesnt check the possible
`KUBVIRT_WITH_CNAO_SKIP_CONFIG`
That allows you to install CNAO but wihtout the CNAO CR (where you say which components to install)

So if CNAO will be installed without CR (i.e using `KUBVIRT_WITH_CNAO_SKIP_CONFIG`)
+ Istio will be installed
and later the CNAO CR will install multus, this WA code wont run and we will have the bug

oshoval replied:
I am fine to not handle it but please don't remove it 
We are using KUBVIRT_WITH_CNAO_SKIP_CONFIG on KubeSecondaryDNS
Its helpful once you want to deploy the cached CNAO, but with your CR
This way you don't need to care about CNAO bumps, as it is part of the kubevirtci TAG that you need to bump anyhow

maiqueb commented:
OK seems it managed to install the kernel modules.

@dhiller your `/approve` was [not handled by the bot](https://github.com/kubevirt/kubevirtci/pull/932#issuecomment-1351251997). Is something wrong with it ?

ormergi commented:
/test check-provision-k8s-1.25
